### PR TITLE
DecodingResult in from_big_compact

### PR DESCRIPTION
**Pull request by yperbasis**

### CODE REVIEW

File: core/silkworm/common/endian.hpp [hunk 1]
```diff
@@ -185,31 +186,33 @@ ByteView to_big_compact(uint64_t value);
...
+//! \brief Parses unsigned integer from a compacted big endian byte form.
+//! \param [in] data : byte view of compacted value. Length must be <= sizeof(UnsignedInteger);
+//! otherwise kOverflow is returned.
+//! \param [out] out: the corresponding integer with native endianness.
+//! \return kOk or kOverflow or kLeadingZero.
+//! \remarks A "compact" big endian form strips leftmost bytes valued to zero;
+//! if the input is not compact kLeadingZero is returned.
 template <typename UnsignedInteger>
-static std::optional<UnsignedInteger> from_big_compact(ByteView data, bool allow_leading_zeros = false) {
+static DecodingResult from_big_compact(ByteView data, UnsignedInteger& out) {
```
Comment (AndreaLanfranchi) [inline core/silkworm/common/endian.hpp:197]: Why remove the possibility to "allow_leading_zeroes" ? I can't recall the rationale but I guess there was a reason for it to be there.

Comment (AndreaLanfranchi) [reply core/silkworm/common/endian.hpp:197]: Also shouldn't we specialize the template ?
```
template <typename UnsignedInteger,
          typename std::enable_if<std::is_integral<UnsignedInteger>::value && std::is_unsigned<UnsignedInteger>::value, T>::type* = nullptr>
```

File: core/silkworm/common/endian_test.cpp [hunk 2]
```diff
@@ -121,8 +121,9 @@ TEST_CASE("Block as key and compact form") {
         auto block_number_compact_bytes{to_big_compact(block_number)};
         CHECK(to_hex(block_number_compact_bytes) == "5485ffde");
         // Convert back and check
-        auto block_number_from_compact{from_big_compact<uint64_t>(block_number_compact_bytes)};
-        CHECK(block_number_from_compact == block_number);
+        uint64_t out64;
```
Comment (AndreaLanfranchi) [inline core/silkworm/common/endian_test.cpp:124]: Shouldn't be initialized regardless ?

File: core/silkworm/common/endian.hpp [hunk 3]
```diff
@@ -185,31 +186,33 @@ ByteView to_big_compact(uint64_t value);
...
     if (data.length() > sizeof(UnsignedInteger)) {
-        return std::nullopt;
+        return DecodingResult::kOverflow;
     }
 
-    UnsignedInteger x{0};
-
+    out = 0;
```
Comment (AndreaLanfranchi) [inline core/silkworm/common/endian.hpp:202]: Imho this is questionable ... should be duty of higher caller to ensure out is validly initialized.
Any error condition should NOT touch out.

File: core/silkworm/common/endian.hpp (hunk 1 above)
Comment (yperbasis) [reply core/silkworm/common/endian.hpp:197]: Unfortunately `is_integral` is not true for `intx`. Perhaps @chfast can enable it.

Comment (yperbasis) [reply core/silkworm/common/endian.hpp:197]: > Why remove the possibility to "allow_leading_zeroes" ? I can't recall the rationale but I guess there was a reason for it to be there.

I dropped allow_leading_zeroes for simplicity. It was used only in one place in cmd/consensus, where I replaced it with `std::stoull` base=16).

File: core/silkworm/common/endian_test.cpp (hunk 2 above)
Comment (yperbasis) [reply core/silkworm/common/endian_test.cpp]: Done in https://github.com/torquem-ch/silkworm/pull/582/commits/d0892f5083faca4051ef738e2d3876a928f7e602

File: core/silkworm/common/endian.hpp (hunk 1 above)
Comment (chfast) [reply core/silkworm/common/endian.hpp:197]: > Unfortunately `is_integral` is not true for `intx`. Perhaps @chfast can enable it.

The `std::is_intergral` is not user-extendable. https://en.cppreference.com/w/cpp/types/is_integral

**Review by AndreaLanfranchi — APPROVED** [review]

File: core/silkworm/common/endian.hpp [hunk 4]
```diff
@@ -185,31 +186,30 @@ ByteView to_big_compact(uint64_t value);
...
-    if (data.empty()) {
-        return x;
+    if (!data.empty() && data[0] == 0) {
+        return DecodingResult::kLeadingZero;
     }
 
-    if (data[0] == 0 && !allow_leading_zeros) {
-        return std::nullopt;
-    }
+    out = 0;
```
Comment (AndreaLanfranchi) [inline core/silkworm/common/endian.hpp:206]: Just a note ... This sets to zero all bytes of out.
Probably a `memset` to zero for all bytes NOT overwritten by subsequent `memcpy` depicts better the logic of the function but I understand it adds some overhead in computing.

**Pull request merged**